graphql_compiler/compiler/ir_lowering_match.py

```diff
@@ -523,6 +527,76 @@ def prune_output_blocks_in_compound_match_query(compound_match_query):
         return CompoundMatchQuery(match_queries=match_queries)
 
 
+def collect_filters_to_first_location_instance(compound_match_query):
+    """Collapse all filters applied to a particular location into a single `as_block`.
+
+    Specifically, for any location `L`, find all filters present on `L`, and apply
+    the conjunction of all these filters to the first occurence of `L` (within its MatchQuery).
+    """
+    def predicate_list_to_where_block(predicate_list):
+        """Convert a list of predicates to an Expression that is the conjunction of all of them."""
+        if not isinstance(predicate_list, list):
+            raise AssertionError(u'Expected `list`, Received {}.'.format(predicate_list))
+        if not predicate_list:
+            return None
+
+        if not isinstance(predicate_list[0], Expression):
+            raise AssertionError(u'Non-predicate object {} found in predicate_list'
+                                 .format(predicate_list[0]))
+        if len(predicate_list) == 1:
+            return predicate_list[0]
+        else:
+            return BinaryComposition(u'&&',
+                                     predicate_list_to_where_block(predicate_list[1:]),
+                                     predicate_list[0])
+
+    new_match_queries = []
+    # Each MatchQuery is processed independently
+    for match_query in compound_match_query.match_queries:
+        location_to_predicates = {}
```

Comment:
this seems to be keyed on `MarkLocation` and not on `Location` -- if you make it use `Location` then `MarkLocation` won't have to be hashable